src/compute/src/render/join/linear_join.rs

```diff
@@ -290,16 +301,186 @@ where
     }
 }
 
+/// Dispatches valid combinations of arrangements where the type-specialized keys match.
+fn dispatch_join_local_local<G>(
+    join_impl: LinearJoinImpl,
+    prev_keyed: SpecializedArrangement<G>,
+    next_input: SpecializedArrangement<G>,
+    closure: JoinClosure,
+) -> (
+    Collection<G, Row, Diff>,
+    Option<Collection<G, DataflowError, Diff>>,
+)
+where
+    G: Scope,
+    G::Timestamp: Lattice,
+{
+    match (prev_keyed, next_input) {
+        (
+            SpecializedArrangement::Bytes9Row(prev_key_types, prev_keyed),
+            SpecializedArrangement::Bytes9Row(next_key_types, next_input),
+        ) => {
+            assert!(prev_key_types
+                .iter()
+                .zip(next_key_types.iter())
+                .all(|(c1, c2)| c1.scalar_type == c2.scalar_type));
+            differential_join_inner(
+                join_impl,
+                prev_keyed,
+                next_input,
+                &prev_key_types,
+                &[],
+                &[],
+                closure,
+            )
+        }
+        (
+            SpecializedArrangement::RowRow(prev_keyed),
+            SpecializedArrangement::RowRow(next_input),
+        ) => differential_join_inner(join_impl, prev_keyed, next_input, &[], &[], &[], closure),
+        _ => panic!("Invalid combination of type specializations: key types differ!"),
```

Comment:
Please mention the types in the panic to make debugging easier.

Reply:
Done!